ipfx/qc_feature_extractor.py

```diff
@@ -59,9 +100,35 @@ def extract_electrode_0(data_set, tags):
 
         e0 = qcf.measure_electrode_0(bath_data.i, bath_data.sampling_rate)
 
-    except IndexError as e:
-        tags.append("Electrode 0 is not available")
-        e0 = None
+    except IndexError as e:  # ontology.bath_names did not find any bath sweeps
+        try:
+            # collect all smoke test sweeps and the first cell attached sweep
+            smoke_test_sweeps = data_set.get_sweep_numbers(ontology.smoketest_names)
+            cell_attached_sweep = data_set.get_sweep_numbers(ontology.seal_names)[0]
+            
+            # find only smoketest sweeps that were ran before cell attached sweep
+            inbath_candidates = filter_smoketests(smoke_test_sweeps, cell_attached_sweep)
+
+            # check pipette resistance of inbath_candidates sweeps and append to 
+            # e0_candidates if pipette resistance is less than 10 MOhms. 
+            e0_candidates = []
+            for sweep in inbath_candidates:
+                bath_data = data_set.sweep(sweep)
+                seal = qcf.measure_seal(bath_data.v, bath_data.i, bath_data.t)
+                print(seal)
+                if abs(seal) < 0.01:
```

Comment:
How do you determine this threshold? Can it be set in terms of the existing qc parameters? What would happen if we did not include this condition here?

Reply:
Up to this point, we did not have ad-hoc thresholds in the code and I would prefer us keeping it that way. Thus if you can think of alternative to setting a threshold 0.01 that would be preferable.

Reply:
@sgratiy I will look into this and see if there is a way that we can correct the data rather than modifying IPFX. The threshold is there because the EXTPSMOKET sweep is usually run with the pipette out of the bath, but for a period of time it was also being run once the pipette was in the bath in place of the EXTPINBATH. We need a way to determine if the pipette was in the bath during the EXTPSMOKET before using it to measure the electrode_0 value, and a low resistance seal (10 MOhm, or 0.01 Ohm) is the only way I can see to do that. The resistance of an open pipette tip in the bath is usually around 3-8 MOhms and will never exceed 10 MOhms (you simply can't patch with a pipette that has a resistance higher than 10 MOhms), and calculating a seal from a smoke test where the pipette is out of the bath will result in a very high seal (usually several GOhms). This is the best way I could figure out how to find EXTPSMOKET sweeps where the pipette was in the bath rather than out of the bath. I will give an update here once I figure out if there is an alternative route we can go here.

Reply:
@sgratiy I see that the white paper states that "Thick walled borosilicate glass (...) electrodes were manufactured (...) with a resistance of 3 to 7 MOhms". Would it be acceptable if I made this threshold a constant and made the comparison to that (instead of the arbitrary 0.01 Ohm threshold defined in the middle of the code)? 

Also, while I see where you are coming from about fixing the data rather than changing the code that processes it, I can see that the extract_electrode_0 function isn't entirely robust as it is, in that it doesn't check the resistance of the pipette to ensure that it is in the range mentioned in the white paper. I think that the resistance check I am doing needs to also be applied even when getting the sweep with the proper in-bath name, EXTPINBATH. I could add another function to check pipette resistance to qc_features.py and this could be called in extract_electrode_0 to ensure that the resistance of the pipette is in range before getting the baseline current (e0). As it is right now, any sweep that is ran with the EXTPINBATH stimulus will have the baseline current extracted even if the pipette was not in the bath, or in some other configuration, and if that baseline current is within range then electrode_0 will pass QC. 

The Ephys team's thoughts on this are that the data are immutable and that the purpose of ipfx is to make use of the data and that it should be able to handle certain errors in the data. I think that it is reasonable for ipfx to check through the smoketest sweeps, EXTPSMOKET, when no sweep with the in-bath name, EXTPINBATH, is found, and see if one of those sweeps was collected with the pipette in the bath, especially because these stimulus waves are exactly the same. The EXTPSMOKET and the EXTPINBATH stimulus waves are exactly the same, just named differently to more easily identify where in the experiment the user should be; however it should not be assumed the user always runs the appropriate stimulus.

I don't think that what I am proposing here deviates from the white paper at all but let me know what you think Sergey, thanks!